### ISSUE TITLE
Non-NPC followers (ie horses, summons from the Arcana mod) getting stuck in basements

### ISSUE THREAD

**Issue by Proxiehunter** [cataclysmbnteam/Cataclysm-BN#1422]
I've seen this happen in let's plays with horses following the player into a basement and refusing to go back up the stairs and today I had a similar issue with a Hunting Horror and a Migo I summoned through the Arcana mod. While I proposed a fix for summons on the Arcana mod github that doesn't really work for horses (and I assume the issue extends to cows, chickens, etc. although chickens dogs and cats can be put in a pet carrier and taken back up). Is there a way to fix it so livestock can't get stuck in basements?

**Comment by Proxiehunter** [comment]
Attached the save with my summons stuck in the basement of my evac shelter.

Mods in use: 
[
  "dda",
  "no_npc_food",
  "novitamins",
  "elevated_bridges",
  "Arcana",
  "mutant_insects",
  "mutant_npcs__",
  "deoxymod__",
  "SzQmod",
  "Tankmod_Revived",
  "StatsThroughSkills",
  "stats_through_kills",
  "Nonperishable_Overhaul"
]

[Range.zip](https://github.com/cataclysmbnteam/Cataclysm-BN/files/8374249/Range.zip)

**Comment by SaintCirno9** [comment]
Not sure if DDA has fixed it, I had this problem too.

**Comment by Coolthulhu** [comment]
It's not a trivial problem to solve because of "teleport stairs". Down stairs aren't actually aligned with up stairs in vast majority of cases. Ideally, stairs would have same x/y coords and only differ on z (by +/-1).
Monster AI works in coordinates and thus has problems with this.

I guess it could be solvable by letting more of the critters use NPC-style pathfinding. NPC pathfinding understands teleport stairs.

**Comment by SaintCirno9** [comment]
You can manually fix the problem by editing the map.
Move the up/down stairs to the same xy location.

**Comment by Coolthulhu** [comment]
> You can manually fix the problem by editing the map.

Yes, but there will still be locations with this problem. I thought about introducing a debug option that turns off stair teleporting (for testing) and maybe a brute force mapgen test that searches for all mapgen pairs (specials with up/down stairs, mostly) where stairs are misaligned.
I'd expect it to find a ton of mismatches.

**Comment by Proxiehunter** [comment]
> I guess it could be solvable by letting more of the critters use NPC-style pathfinding. NPC pathfinding understands teleport stairs.

Would giving anything considered "friendly" to the player NPC-style pathfinding be easier or harder than expanding it to multiple critters in a more general way? That's really where the problems are happening, with critters that are friendly to the player and thus follow them down stairs they can't get back up.

**Comment by Proxiehunter** [comment]
Would NPC style pathfinding also help them with raised bridges? I discovered today my summoned Migo won't cross a raised bridge at all, my summoned hunting horror tried and got stuck on the bridge, and I had to ride my Dark Wyrm over the bridge but it wouldn't follow on it's own. I assume horses, cows, and other livestock would do the same. 

If you want to bring more than one horse/cow over a bridge and don't have a vehicle you put them in and drive over the bridge yet riding them over one at a time would be a pain in the ass. Especially since I often find my characters too heavy to ride horses unless I dump pretty much _all_ of my equipment. Which has lead to me selecting minimum height for all of my characters as that seems to make them light enough to carry more equipment and still ride a horse if I find it. Might raise the fact that you pretty much be the size of a typical jockey to ride a horse as a separate issue later.

**Comment by chaosvolt** [comment]
Side note, as I mused in the relevant issue, porting over https://github.com/CleverRaven/Cataclysm-DDA/pull/50890 should make this less of a problem since that'd give players a manual solution, only issue is I dunno if that PR tested trying to lead up/down stairs or not.

Oh wait, okay.

> Works across z-levels as long as the up/down stairs are aligned (otherwise the player will be too far from the animal after using the stairs)

Well shit.

**Comment by LyleSY** [comment]
This is a big issue for DinoMod also. I have a horrible hack I'm planning to port over but an actual fix would be even better

**Comment by chaosvolt** [comment]
Anything helps hopefully. On Arcana's end, I have set aside https://github.com/chaosvolt/cdda-arcana-mod/pull/275 as a thing that will help make summoned monsters less annoying to wrangle, but there's some UI stuff I want to make fuck with beforehand...

**Comment by chaosvolt** [comment]
The above Arcana PR is now set so that once https://github.com/cataclysmbnteam/Cataclysm-BN/pull/1931 I should be good to merge that, so that'll take care of things on Arcana's end at least.

https://github.com/cataclysmbnteam/Cataclysm-BN/pull/1905 should also improve this once merged, via adding an additional way to get an ornery monster follower up and down stairs (I hope, I might want to check out their branch to see for myself).
